parlai/agents/seq2seq/modules.py

```diff
@@ -278,13 +281,39 @@ def forward(self, xs, ys=None, cand_params=None, prev_enc=None,
         return scores, cand_scores, encoder_states
 
 
+class UnknownDropout(nn.Module):
+    """With set frequency, replaces tokens with unknown token.
+
+    This layer can be used right before an embedding layer to make the model
+    more robust to unknown words at test time.
+    """
+
+    def __init__(self, unknown_idx, probability):
+        """Initialize layer.
+
+        :param unknown_idx: index of unknown token, replace tokens with this
+        :param probability: during training, replaces tokens with unknown token
+                            at this rate.
+        """
+        super().__init__()
+        self.unknown_idx = unknown_idx
+        self.prob = probability
+
+    def forward(self, input):
+        """If training and dropout rate > 0, masks input with unknown token."""
+        if self.training and self.prob > 0:
+            mask = input.new(input.size()).float().uniform_(0, 1) < self.prob
+            input.masked_fill_(mask, self.unknown_idx)
```

Comment:
Did you happen to check whether the destructive fill could have problems here? I just wonder, if `text_vec` is cached (as in `-bsrt True`) that the `text_vec` could be thrashed and filled with UNKs.

Reason this shouldn't happen: We're applying this to the output of `padded_tensor`, which produces a new tensor every iteration, and just copies over from the (possibly cached) `text_vec`.